File: day15/Day15.scala
import scala.io.Source.stdin
import scala.collection.mutable.{Map as MutMap, Set as MutSet, PriorityQueue}

type Point = (Int, Int)

@main def main(idx: Int) =
  var xs = stdin.getLines().zipWithIndex.flatMap{
             case (l, y) => l.map{_.asDigit}.zipWithIndex.map{
               case (v, x) => ((y, x), v) }}.toMap
  def boundBy(f: Point => Int) = xs.keys.map(f).max + 1
  var h = boundBy(_._1)
  var w = boundBy(_._2)

  if (idx == 2) then
    xs = xs.flatMap{
      case ((y, x), v) =>
         for { sy <- 0 until 5; sx <- 0 until 5
               tv = v + sy + sx
               nv = if (tv <= 9) tv else tv - 9
         } yield ((y + sy * h, x + sx * w), nv)
      }
    h *= 5
    w *= 5

  val dist = xs.view.mapValues{_ => Int.MaxValue}.to(MutMap)
  val seen = MutSet[Point]()
  val q = new PriorityQueue(using Ordering[(Int, Point)].reverse)

  dist((0, 0)) = 0
  q.enqueue((0, (0, 0)))

  val neighbors = List((-1, 0), (0, -1), (0, 1), (1, 0))
  while !q.isEmpty do
    val (len, (y, x)) = q.dequeue()
    for { (dy, dx) <- neighbors
          np@(ny, nx) = (y + dy, x + dx)
          if 0 <= ny && ny < h && 0 <= nx && nx < w
    } do
      val old = dist(np)
      val new_ = len + xs(np)
      if new_ < old then
        dist(np) = new_
        q.enqueue((new_, np))

  println(dist((h - 1, w - 1)))

File: day15/Day15.rs
use std::cmp::Reverse;
use std::collections::BinaryHeap;
use std::collections::HashMap;
use std::collections::HashSet;
use std::env;
use std::io;

fn main() {
    type Point = (i32, i32);
    let nums: HashMap<Point, u32> = io::stdin()
        .lines()
        .map(Result::unwrap)
        .enumerate()
        .flat_map(|(y, l)| {
            l.chars()
                .enumerate()
                .map(|(x, c)| ((y as i32, x as i32), c.to_digit(10).unwrap()))
                .collect::<Vec<_>>()
        })
        .collect();

    let bound_by = |f: fn(&Point) -> i32| nums.keys().map(f).max().unwrap() + 1;
    let h = bound_by(|p| p.0);
    let w = bound_by(|p| p.1);

    let (h, w, nums) = if env::args().nth(1).unwrap() == "1" {
        (h, w, nums)
    } else {
        (
            h * 5,
            w * 5,
            nums.iter()
                .flat_map(|((y, x), v)| {
                    (0..5).flat_map(move |sy| {
                        (0..5).map(move |sx| {
                            let nv = match *v + sy as u32 + sx as u32 {
                                x if x <= 9 => x,
                                x => x - 9,
                            };

                            ((*y + sy * h, *x + sx * h), nv)
                        })
                    })
                })
                .collect(),
        )
    };

    let mut dist: HashMap<Point, u32> = HashMap::new();
    let mut q: BinaryHeap<Reverse<(u32, Point)>> = BinaryHeap::new();
    let mut seen: HashSet<Point> = HashSet::new();

    dist.insert((0, 0), 0);
    q.push(Reverse((0, (0, 0))));

    while let Some(Reverse((len, p @ (y, x)))) = q.pop() {
        for (dy, dx) in [(-1, 0), (0, -1), (0, 1), (1, 0)] {
            let np @ (ny, nx) = (y + dy, x + dx);

            if 0 <= ny && ny < h && 0 <= nx && nx < w {
                let &old = dist.get(&np).unwrap_or(&u32::MAX);
                let new = len + nums.get(&np).unwrap();

                if new < old {
                    dist.insert(np, new);
                    q.push(Reverse((new, np)));
                }
            }
        }

        seen.insert(p);
    }

    println!("{}", dist[&(h - 1, w - 1)]);
}
